src/chart/funnel/funnelLayout.ts

```diff
@@ -337,55 +427,156 @@ export default function funnelLayout(ecModel: GlobalModel, api: ExtensionAPI) {
+                    let end;
+
+                    if (index === indices.length - 1 && exitShape === 'rect') {
+                        end = getLinePointsBySize(pos + sideLen, resSize / maxSize * viewSize);
+                    }
+                    else {
+                        resSize += sort === 'ascending' ? sideLen : -sideLen;
+                        end = getLinePointsBySize(pos + sideLen, resSize / maxSize * viewSize);
                     }
+
+                    data.setItemLayout(idx, {
+                        points: start.concat(end.slice().reverse())
+                    });
+                    return;
                 }
+                else if (showRate) {
+                    const dataStart = getLinePoints(idx, pos);
+                    let dataEnd;
+                    if (exitWidth !== undefined && index === indices.length - 1) {
+                        const val = data.get(valueDim, idx) as number || 0;
+                        const itemSize = linearMap(val, [min, max], sizeExtent, true);
+                        const exitSize = itemSize * (exitWidth as number) / 100;
+                        dataEnd = getLinePointsBySize(pos + sideLen / 2, exitSize);
+                    }
+                    else {
+                        dataEnd = getLinePoints(idx, pos + sideLen / 2);
+                    }
+                    const rateStart = getLinePoints(idx, pos + sideLen / 2);
+                    const rateEnd = getLinePoints(nextIdx, pos + sideLen);
 
-                const start = getLinePoints(idx, y);
-                const end = getLinePoints(nextIdx, y + height);
+                    const val = data.get(valueDim, idx) as number || 0;
+                    const nextVal = data.get(valueDim, nextIdx) as number || 0;
+                    let rate: number | string = nextVal / val;
+                    rate = 'Rate ' + (rate * 100).toFixed(0) + '%';
+                    if (index === 0) {
+                        firstVal = val;
+                    }
+                    else if (index === indices.length - 1) {
+                        const lastVal = val;
+                        rate = lastVal / firstVal;
+                        rate = 'Overall rate ' + (rate * 100).toFixed(0) + '%';
+                    }
 
-                y += height + gap;
+                    data.setItemLayout(idx, {
+                        points: dataStart.concat(dataEnd.slice().reverse()),
+                        ratePoints: rateStart.concat(rateEnd.slice().reverse()),
+                        isLastPiece: index === indices.length - 1,
+                        rate
+                    });
+                    return;
+                }
+                const start = getLinePoints(idx, pos);
+                let end;
+
+                if (index === indices.length - 1 && exitShape === 'rect') {
+                    end = getLinePoints(idx, pos + sideLen);
```

Comment:
Like the comments above. The code can be tighter here.
```ts
end = getLinePoints( index === indices.length - 1 && exitShape === 'rect' ? idx : nextIdx, pos + sideLen);
```

Reply:
Ok, I will fix code styles what you say 😁